sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/datasketchesAggregates.scala

```diff
@@ -0,0 +1,523 @@
+
+  override def withNewMutableAggBufferOffset(newMutableAggBufferOffset: Int):
+  HllSketchUnionEstimate = copy(mutableAggBufferOffset = newMutableAggBufferOffset)
+
+  override def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): HllSketchUnionEstimate =
+    copy(inputAggBufferOffset = newInputAggBufferOffset)
+
+  override protected def withNewChildrenInternal(newLeft: Expression, newRight: Expression):
+  HllSketchUnionEstimate = copy(child = newLeft, lgMaxKExpression = newRight)
+
+  // overrides for BinaryLike
+
+  override def left: Expression = child
+  override def right: Expression = lgMaxKExpression
+
+  // Type checking
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    (left.dataType, right.dataType) match {
+      case (_, NullType) =>
+        DataTypeMismatch(
+          errorSubClass = "UNEXPECTED_NULL",
+          messageParameters = Map(
+            "exprName" -> "lgMaxK"
+          )
+        )
+      case (BinaryType, IntegerType) =>
+        if (!right.foldable) {
+          DataTypeMismatch(
+            errorSubClass = "NON_FOLDABLE_INPUT",
+            messageParameters = Map(
+              "inputName" -> "lgMaxK",
+              "inputType" -> toSQLType(right.dataType),
+              "inputExpr" -> toSQLExpr(right)
+            )
+          )
+        } else if (lgMaxK <= 0L) {
+          DataTypeMismatch(
+            errorSubClass = "VALUE_OUT_OF_RANGE",
+            messageParameters = Map(
+              "exprName" -> "lgMaxK",
+              "valueRange" -> s"[0, positive]",
+              "currentValue" -> toSQLValue(lgMaxK, IntegerType)
+            )
+          )
+        } else {
+          TypeCheckSuccess
+        }
+      case _ =>
+        DataTypeMismatch(
+          errorSubClass = "HLLSKETCH_WRONG_TYPE",
+          messageParameters = Map(
+            "functionName" -> toSQLId(prettyName),
+            "expectedRight" -> toSQLType(IntegerType),
+            "actual" -> Seq(left.dataType, right.dataType)
+              .map(toSQLType).mkString(", ")
+          )
+        )
+    }
+  }
+
+  // Implementations specific to HllSketchUnionEstimate
+
+  override def prettyName: String = "hllsketch_union_estimate"
+
+  override def dataType: DataType = LongType
+
+  override def nullable: Boolean = false
+
+  /**
+   * Instantiate an Union instance using the lgMaxK param.
+   *
+   * @return an Union instance
+   */
+  override def createAggregationBuffer(): Union = {
+    new Union(lgMaxK)
+  }
+
+  /**
+   * Update the Union instance with the HllSketch byte array obtained from the row.
+   *
+   * @param union The Union instance.
+   * @param input an input row
+   */
+  override def update(union: Union, input: InternalRow): Union = {
+    val v = child.eval(input)
+    if (v != null) {
+      child.dataType match {
+        case BinaryType =>
+          union.update(HllSketch.wrap(WritableMemory.writableWrap(v.asInstanceOf[Array[Byte]])))
```

Comment:
Why do we use `WritableMemory` here?  Wouldn't it be enough to use `Memory` (immutable memory wrapper) instead?

Reply:
Good catch - will update.